### ISSUE TITLE
[A2-40] 整网融合算子：清单、ABI 与验收设计（主机侧）

### ISSUE THREAD

**Issue by limjiunnbin** [ddddwee1/ascend_fla_dev#54]
**A2-40** · 波次 `W0` · SoC `any` · 优先级 `P1` · 状态 `done`

| 依赖 | 写集 | 需要 |
|---|---|---|
| 无 | `docs/research/a2_fusion_ops.md` | npu=False ascriptor=True fla=True |

| 轴 | 取值 |
|---|---|
| 模型 | `kimi-linear-48b-a3b` |
| kernel | `causal_conv1d`、`fused_rms_norm_gated`、`qk_l2norm_gate`、`packed_projection` |
| 数据类型 | `bf16`、`fp32` |
| 机器 | `any` |

**本条就是 `causal_conv1d` 链的起点。**
　链上顺序：#54（A2-40） → #55（A2-41）
**本条就是 `fused_rms_norm_gated` 链的起点。**
　链上顺序：#54（A2-40） → #55（A2-41）
**本条就是 `qk_l2norm_gate` 链的起点。**
　链上顺序：#54（A2-40） → #56（A2-42）
**本条就是 `packed_projection` 链的起点。**
　链上顺序：#54（A2-40） → #57（A2-43）

**申领**：先读 `AGENTS.md`、`docs/pm/PROTOCOL.md`、`docs/pm/START.md`，再在本 issue 下评论 `[FLA-PM] APPLY A2-40 from=<你的 GitHub 账号>`（模板见 PROTOCOL §3.1）。只认 @limjiunnbin 发出的 ASSIGN / REVIEW / CLOSE。

---

# A2-40 整网融合算子：清单、ABI 与验收设计（主机侧）

- 波次 / SoC：W0 / any。纯主机侧调研 + 设计，不需要真机
- 优先级：P1 · 时限：12h · 依赖：无
- 需要：ascriptor workspace（查可复用的算法单元）、fla（对照它把什么放进了 kernel）
- 写集：`docs/research/a2_fusion_ops.md`

## 为什么要有这条

算子本身跑得快，不等于整网跑得快。实测（`docs/handoff.md`）：KDA 整层 decode 一步 458µs，
**KDA 算子只占 18%，层里其余那十几个小算子占 67%** —— 七个投影、三个短卷积、两次 fp32 l2norm 往返、
一个 output norm，每个几乎没有计算量却各要一次 launch。把 kernel 再快一倍，整层只快 9%。

而且这些"周边"现在全是 torch 原生算子，带来两个后果：缺内置算子包的机器上层级验证根本做不了；
报层级性能时得把它们拆出来说，否则等于把 torch 的开销记在本仓算子账上。

本任务不写 kernel，只把**要做哪些、ABI 长什么样、怎么验收**定下来，供后续 A2-41 / A2-42 / A2-43 执行。

## 四个对象（看板 fusion 族）

| id | 现状 | 对应缺口 |
|---|---|---|
| `causal_conv1d` | `modules/convolution.py`，torch 实现，支持 cache 单步解码与 `T < kernel_size` 补零 | `modules-are-torch-not-kernels` |
| `fused_rms_norm_gated` | `modules/fused_norm_gated.py`，KDA 的 o_norm，sigmoid 门控 | 同上 |
| `qk_l2norm_gate` | 现在由层做：q/k 的 L2 归一化、门控变换、beta sigmoid | `qk-l2norm-not-in-kernel`（**P1，本仓唯一"错了不报错"的语义缺口**） |
| `packed_projection` | 七个投影分开算；合投影与图捕获都还没做 | `decode-layer-overhead` |

## 步骤

1. 逐个对照 fla：它把哪几步放进了 kernel（`use_qk_l2norm_in_kernel` 等开关），我们现在放在哪、为什么。
   给出语义差异表，标明哪些是"数值上合法但结果静默错"的面。
2. 查 ascriptor 侧可复用的算法单元：`matrix_normalization` 的 `row_l2`、`gated_approximations` 等，
   逐个写清能用哪一部分、缺哪一部分。`docs/matrix/ops.json` 的 `reusable_primitives` 是起点。
3. 为每个对象写 ABI 草案：输入输出张量、dtype、布局、标量、`block_dim` 约束、以及不满足时报什么错
   （`AGENTS.md` §7：不满足即报错，绝不静默兜底）。
4. 写验收判据：对哪个 oracle、什么形状、什么阈值。fp32 判定，报相对 L2 / max_abs_diff。
   `qk_l2norm_gate` 要特别说明怎么证明"融进 kernel 之后与层外做的逐位或按预算一致"。
5. 估工作量与拆分建议：哪些能进 A2-41（独立单元）、哪些必须进 kernel 批次（A2-42 改前向 kernel）、
   哪些是纯层侧改动（A2-43）。
6. 量程逐处判定（`AGENTS.md` §6）：这四个对象里所有 `exp` / 除法 / 规约类算式列表判量程，
   不要只看报错的那一处。

## 验收

- [ ] 四个对象每个都有：现状、与 fla 的语义差异、ABI 草案、验收判据、工作量估计。
- [ ] 可复用的 ascriptor 单元逐个点名，缺的部分写清楚。
- [ ] 量程表覆盖四个对象的全部相关算式，逐处给结论（含"恒 ≤1、下溢到 0 是正确结果"这类判断）。
- [ ] 拆分建议能直接落到看板（A2-41 / A2-42 / A2-43 的范围与依赖）。
- [ ] 没有写或改任何 kernel、模块代码。

## 已知陷阱

- `decode-layer-overhead` 那条缺口里写着处置顺序：图捕获 → 合投影 → 去 l2norm 的 dtype 往返。
  设计时要说明为什么按这个顺序，或者给出更好的理由推翻它。
- **profile 之前不要相信任何性能推断**（`AGENTS.md` §6 铁律一）。本任务不跑真机，所以所有性能话术
  都要标成"待测"，不能写成结论。
- `ShortConvolution` 本来就支持 cache 单步解码，别把已有能力当成缺口重做一遍。

**Comment by joshjms** [comment]
[FLA-PM] APPLY A2-40 from=joshjms
agent: Claude Opus 5 (Claude Code)
socs: none
soc_details: none
ascriptor: yes library 90cfcdc / kernels b3b3f9c（agent/compatibility.json 当前 pin，按 AGENTS.md §3 的 gitcode 来源取得；本地 checkout 未改动，以 git archive 解出件为准）
fla: yes 0.5.2
python: 3.11.16 / torch 2.14.0+cpu / pytest 可用
fork: joshjms/ascend_fla_dev
availability: 异步
note: |
  纯主机侧调研 + 设计（npu=false），与本机能力匹配；手上无进行中的任务（A2-04 已 done）。
  已读 docs/pm/tasks/A2-40.md 与其引用：modules/convolution.py、modules/fused_norm_gated.py、layers/kda.py，
  gaps.json 的 modules-are-torch-not-kernels / qk-l2norm-not-in-kernel / decode-layer-overhead，ops.json 的 reusable_primitives。
  只写 docs/research/a2_fusion_ops.md；不写或改任何 kernel、模块代码，不碰 reserved_paths（GDN-2 的模型专用单元只作为先例引用）。
  handoff.md 里 decode 的 458µs / 18% / 67% 是 A5 实测，报告里会按 SoC 标注并把所有性能话术标成"待测"。

**Comment by limjiunnbin** [comment]
```
[FLA-PM] ASSIGN A2-40 from=limjiunnbin
assignee: joshjms
spec: docs/pm/tasks/A2-40.md
branch: task/A2-40
timebox_h: 12
ack_within_h: 24
report_every_h: 24
```

能力匹配（`needs.npu=false`，ascriptor + fla 都有），写集 `docs/research/a2_fusion_ops.md`
与在飞任务无重叠，你手上无其它任务（A2-04 已 done）。

两点：

1. **只写研究文档，不动任何 kernel / 模块代码**——你自己的 APPLY 说明里已经写清楚了，
   我确认一下没有歧义。GDN-2 的模型专用单元你说"只作为先例引用"，可以读，别写。
2. **性能话术全部标"待测"**——`handoff.md` 那组 458µs/18%/67% 是 A5 实测数，你的产出是主机侧
   设计文档，不要把它们当成本任务自己的结论重复一遍而不加限定。

按 §3.3 24 小时内回 ACK。

**Comment by joshjms** [comment]
[FLA-PM] ACK A2-40 from=joshjms
read: AGENTS.md, docs/handoff.md, docs/pm/PROTOCOL.md, docs/pm/tasks/A2-40.md
plan: |
  1. 对照 fla 0.5.2（kda/layer + ops）逐个列四个对象的现状与语义差异，标出"静默错"面
  2. 在 ascriptor pin（library 90cfcdc / kernels b3b3f9c）里点名可复用单元，写清缺口
  3. 每个对象写 ABI 草案（张量/dtype/布局/标量/block_dim/报错）+ fp32 验收判据
  4. 量程表覆盖四个对象全部 exp/除法/规约；拆分建议落到 A2-41/42/43
  5. 只写 docs/research/a2_fusion_ops.md，性能话术全部标"待测"，A5 数字按 SoC 标注
eta_h: 8
note: ACK 晚于 24h 窗口约 10 分钟；若已退回 open 请告知，按 PM 处置。

**Comment by joshjms** [comment]
[FLA-PM] DONE A2-40 from=joshjms
pr: #85
commits: 4fd8baf
acceptance: |
  - [x] 四个对象每个都有现状、与 fla 的语义差异、ABI 草案、验收判据、工作量估计 —— §1 总览、§2 差异表 D1–D11、§4.1–4.4、§5.1–5.4、§6
  - [x] 可复用 ascriptor 单元逐个点名并写清缺口 —— §3，共 7 个上游单元和 4 个本仓先例（pin 为 kernels b3b3f9c / library 90cfcdc）
  - [x] 量程表覆盖四个对象的全部相关算式，逐处给结论 —— §8 共 17 处：4 处要求特定写法（exp 上溢方向），2 处要求 eps 不能省，1 处需要装闸（门控跨度，已有的闸换实现位置），其余是"实际不可达"或"下溢到 0 就正确"
  - [x] 拆分建议能直接落到看板 —— §6：A2-41/42/43 的范围、依赖、gate、工时，另新提 A2-44（纯主机侧，无 gate）
  - [x] 没有写或改任何 kernel 或模块代码 —— diff 只有 docs/research/a2_fusion_ops.md
host_tests: 未运行 pytest（纯文档改动，本机未建 .venv）；tools/gen_matrix.py --check → "docs/matrix/README.md 是最新的"
device: none
numbers: |
  无真机数字。唯一的算例是解析计算：‖x‖=1e-4 时，fla 的 l2norm 输出范数为 0.0995，本仓 F.normalize 为 1.0（§2 D1）
evidence: |
  fla v0.5.2（9c8e42e）：modules/l2norm.py:43,105；ops/kda/fused_recurrent.py:155；ops/kda/chunk.py:56-62,166-170；
  layers/kda.py 的 use_short_conv else 分支（q/k/v 过 F.silu）；modules/conv/short_conv.py 的 step（原地更新 cache）
  ascriptor library 90cfcdc：docs/cce-support.md:345（c310 是寄存器级向量）；kernels b3b3f9c：matrix_normalization README（没有 eps）、swiglu_bf16 kernel（"supports A2/A3"）
matrix_delta: none
handoff_notes: |
  建议 PM 落到 gaps.json（写集不含 matrix，所以只写在这里）：
  1. 新缺口：l2norm eps 公式与 fla 不同（D1，静默，只在近零行上出现），以及 use_short_conv=False 缺 silu（D3，静默）。两者都是层侧改动，归 A2-44
  2. qk-l2norm-not-in-kernel：静默面比现在的描述窄（照抄 fla 的 flags 会 TypeError）；语义修正（A2-44，不动 kernel）和性能融合（A2-42）可以拆开
  3. decode-layer-overhead：evidence 里的形状是 hidden=2048/H16/HV32，不是 Kimi 的 2304/H32/HV32；每层是 9 个 nn.Linear，不是 7 个；处置 ③ 建议并入 A2-42
  需要拍板：纯向量单元是否从 A2-11 总闸放宽（§6），本文默认不放宽

**Comment by limjiunnbin** [comment]
[FLA-PM] CLOSE A2-40 from=limjiunnbin

已合入 main（PR #85 → 1c1abe3）。

PM 独立复核：
- diff 只有 `docs/research/a2_fusion_ops.md`（写集内），未改任何 kernel/模块代码。
- `gen_matrix.py --check` 独立复算通过。
- 四个对象（causal_conv1d/fused_rms_norm_gated/qk_l2norm_gate/packed_projection）的现状、
  与 fla v0.5.2 的语义差异、ABI 草案、fp32 验收判据、工作量估计齐全；量程表 17 处逐条给了结论。

已落到看板/矩阵的后续动作：
- `gaps.json` 新增两条：`kda-layer-l2norm-eps-formula-diverges`（P2，D1）、
  `kda-layer-missing-silu-when-no-short-conv`（P3，D3）；`qk-l2norm-not-in-kernel` 补充了
  静默面收窄的说明，`decode-layer-overhead` 补充了形状/投影数量的更正。
- 新任务 `A2-44`（#86）：纯主机侧修 D1/D3 + 加 fla 式原始输入 API + 定义域检查，不动 kernel，
  无 gate，可以直接 APPLY。
- `A2-41`/`A2-42`/`A2-43` 沿用文中的拆分建议，依赖 `A2-40` 均已满足（`A2-42` 仍需
  `kernel-batch-approval`，`A2-41`/`A2-43` 仍需 `machines:a2`）。

有一条文中提出但本次不处理，留给用户/PM 后续拍板：**A2-41 涉及的四个对象都是纯向量运算、
不经过 cube，是否可以从 AGENTS.md §2「A2-11 之前任何算子结论都不算数」这条总闸里放宽**。
这是规则层面的决定，PM 不单方面做，按 §6 默认维持不放宽。
